src/OfflineTool.cpp

```diff
@@ -0,0 +1,473 @@
+    };
+    try
+    {
+        auto tx = make_sttx(boost::trim_copy(data));
+        if (tx)
+        {
+            auto rippleKey = RippleKey::make_RippleKey(keyFile);
+
+            rippleKey.multiSign(*tx);
+
+            std::cout << tx->getJson(0).toStyledString() << std::endl;
+        }
+        else
+            fail();
+    }
+    catch (std::exception const&)
+    {
+        fail();
+    }
+}
+
+void
+doCreateKeyfile(boost::filesystem::path const& keyFile,
+    boost::optional<std::string> const& keytype,
+    boost::optional<std::string> const& seed)
+{
+    using namespace ripple;
+    using namespace serialize;
+
+    if (exists(keyFile))
+        throw std::runtime_error(
+            "Refusing to overwrite existing key file: " +
+            keyFile.string());
+
+    auto const kt = keytype ?
+        boost::optional<KeyType>{keyTypeFromString(*keytype)} :
+        boost::none;
+    if (kt && *kt == KeyType::invalid)
+    {
+        std::cout << "Invalid key type: \"" << *keytype << "\"" <<
+            std::endl;
+        return;
+    }
+
+    auto key = RippleKey::make_RippleKey(kt, seed);
+
+    key.writeToFile(keyFile);
+
+    std::cout << "New ripple key created " <<
+        "in " << keyFile.string() << "\n" <<
+        "Key type is " << to_string(key.keyType()) << ", and " <<
+        "account ID is " <<
+        toBase58(calcAccountID(key.publicKey())) << "\n" <<
+        "\nThis file should be stored securely and not shared\n\n";
+
+}
+
+std::string
+getStdin()
+{
+    std::ostringstream stdinput;
+    stdinput << std::cin.rdbuf();
+    return stdinput.str();
+}
+
+void
+runCommand (const std::string& command,
+    std::vector <std::string> const& args,
+    boost::filesystem::path const& keyFile,
+    boost::optional<std::string> const& keyType,
+    boost::optional<std::string> const& seed)
+{
+    using namespace std;
+
+    struct commandParams
+    {
+        vector<string>::size_type minParams;
+        vector<string>::size_type maxParams;
+        bool enableStdin;
+    };
+    static map<string, commandParams>
+        const commandArgs = {
+            { "serialize", {0, 1, true} },
+            { "deserialize", {0, 1, true} },
+            { "sign", {0, 1, true} },
+            { "multi_sign", {0, 1, true} },
+            { "create_keyfile", {0, 0, false} },
+    };
+
+    auto const iArgs = commandArgs.find(command);
+
+    if (iArgs == commandArgs.end())
+        throw std::runtime_error("Unknown command: " + command);
+
+    if (args.size() < iArgs->second.minParams ||
+            args.size() > iArgs->second.maxParams)
+        throw std::runtime_error("Syntax error: Wrong number of arguments");
+
+    auto const input = [&]
+    {
+        if (args.size() == 0)
+            return iArgs->second.enableStdin ? getStdin() : string{};
+        else
+            return args[0];
+    }();
+
+    if (command == "serialize")
+        doSerialize(input);
```

Comment:
Here's a thought.  Suppose you passed `doSerialize`, `doDeserialize`, `doSingleSign`, `doMultiSign`, and `doCreateKeyfile` all the same parameters.  So you'd be passing the same four parameters (`input`, `keyFile`, `keyType`, and `seed`) to all five functions.  Then you could put a function pointer in `CommandParams` that points to the function you should call.

That would have the advantage that you avoid this `if/else` chain.

It would have the disadvantage that those 5 functions are now tied together with a non-essential commonality.

I'm not absolutely positive it's a win.  But I thought I'd bring it up for consideration.